Add Pod Level Resources Blog Post (#51575)
kubernetes/website: content/en/blog/_posts/2025-xx-xx-pod-level-resources/index.md

---
layout: blog
title: "Pod Level Resources Graduated to Beta"
date: 2025-xx-xx
draft: true
slug: kubernetes-v1-34-pod-level-resources
author: Dixita Narang (Google)
---

> On behalf of the Kubernetes community, I am thrilled to announce that the **Pod Level Resources** feature **has graduated to Beta** in the Kubernetes v1.34 release! This significant milestone introduces a new layer of flexibility for defining and managing resource allocation for your Pods. This flexibility stems from the ability to **specify CPU and memory resources directly at the Pod level**, for the Pod as a whole, in addition to or instead of container-level specifications.

Reviewer: Write it as "Pod level resources", sentence case, and in italic (not bold). This paragraph shouldn't have any boldface text in it.

> #### **What are Pod Level Resources?**

Reviewer: Don't add the boldface here

> Traditionally, resource specifications for Pods were primarily defined at the

Reviewer: ```suggestion
Until recently, resource specifications that applied to Pods were primarily defined at the
```

individual container level. While effective, this approach sometimes required
duplicating or meticulously calculating resource needs across multiple containers

> within a single Pod. Pod Level Resources introduces the ability to specify CPU,

Reviewer: ```suggestion
within a single Pod. As an opt-in, beta feature, Kubernetes allows you to specify the CPU,
```

> memory and hugepages resources directly at the **Pod-level**. This means you can now define

Reviewer: ```suggestion
memory and hugepages resources directly at the Pod-level. This means you can now define
```

resource requests and limits for an entire Pod, providing a consolidated approach to
resource declaration without requiring granular, per-container micromanagement where
it's not needed.

> #### **Why Does Pod Level Resources Matter?**

Reviewer: Don't add the boldface here

> This feature enhances resource management in Kubernetes by offering **flexible resource management** at both the Pod and container levels.

Reviewer: Use italics not bold here.

> * It provides a **consolidated approach** to resource declaration, reducing the need for meticulous, per-container micromanagement, especially for Pods with multiple containers.

Reviewer: No need for bold here.

> * When both pod-level and container-level resources are specified, **pod-level requests and limits take precedence**, offering a powerful way to enforce overall resource ceilings for your Pods.

Reviewer: ```suggestion
* When both pod-level and container-level resources are specified, pod-level requests and limits take precedence.
  This gives you – and cluster administrators - a powerful way to enforce overall resource ceilings for your Pods.
```
Maybe explain this more later with an example of an injected sidecar that doesn't raise the Pod-level CPU limit(?)

> * Pod-level resources are **prioritized** in influencing the Quality of Service (QoS) class of the Pod.

Reviewer: For this detail, link to the place in the docs that explains that.

Author: The doc doesn't explain the details: https://kubernetes.io/docs/tasks/configure-pod-container/assign-pod-level-resources/#:~:text=QoS%3A%20Pod%2D-,level,-resources%20take%20precedence

It just briefly mentions the qos with pod-level resources. The KEP mentions the details - https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#qos-changes

do you think we should add this detail to the documentation?

> * The Out-Of-Memory (OOM) score adjustment calculation **considers both pod-level and container-level resources**.

Reviewer: ```suggestion
* For Pods running on Linux nodes, the Out-Of-Memory (OOM) score adjustment calculation
  considers both pod-level and container-level resources.
```

* Pod-level resources are **designed to be compatible with existing Kubernetes functionalities**, ensuring a smooth integration into your workflows.

> #### **How to Use Pod Level Resources**

Reviewer: Don't add the boldface here

> To utilize Pod Level Resources, the **<code>PodLevelResources</code> feature gate must be explicitly enabled** for your control plane and for all nodes in your cluster. Your Kubernetes server must be at or later than version 1.32 to use this feature.

Reviewer: ```suggestion
To utilize Pod level resources, the `PodLevelResources` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
must be explicitly enabled** for your control plane
and for **all** nodes in your cluster.
Your Kubernetes cluster must be able components at or later than version 1.32 to use this feature.
```

> **Defining Pod-Level Resources:** You can specify CPU, memory and hugepages resources directly in the Pod spec manifest at the `resources` field for the entire Pod.

Reviewer: This bold text should be a heading in its own right?

Here’s an example demonstrating a Pod with both CPU and memory requests and limits
defined at the Pod level:

> ```

Reviewer: append "yaml" here (without the quotes)

apiVersion: v1
kind: Pod
metadata:
  name: pod-resources-demo
  namespace: pod-resources-example
spec:
  resources: # Pod-level resources
    limits:
      cpu: "1"
      memory: "200Mi"
    requests:
      cpu: "1"
      memory: "100Mi"
  containers:
  - name: pod-resources-demo-ctr-1
    image: nginx
    resources:
      limits:
        cpu: "0.5"
        memory: "100Mi"
      requests:
        cpu: "0.5"
        memory: "50Mi"
  - name: pod-resources-demo-ctr-2
    image: fedora
    command: ["sleep", "inf"]
```

> 

Reviewer: Consider using inline comments to explain further.

In this example, the `pod-resources-demo` Pod as a whole requests 1 CPU and 100 MiB of memory, and is limited to 1 CPU and 200 MiB of memory. The containers within will operate under these overall Pod-level constraints, as explained in the next section.

> **Interaction with Container-Level Resources:** When both pod-level and container-level resources are specified, **pod-level requests and limits take precedence**. This means the node allocates resources based on the pod-level specifications.

Reviewer: This should be an actual heading.

Consider a Pod with two containers where pod-level CPU and memory requests and
limits are defined, and only one container has its own explicit resource
definitions:

> ```

Reviewer: append "yaml" here (without the quotes)

apiVersion: v1
kind: Pod
metadata:
  name: pod-resources-demo
  namespace: pod-resources-example
spec:
  resources:
    limits:
      cpu: "1"
      memory: "200Mi"
    requests:
      cpu: "1"
      memory: "100Mi"
  containers:
  - name: pod-resources-demo-ctr-1
    image: nginx
    resources:
      limits:
        cpu: "0.5"
        memory: "100Mi"
      requests:
        cpu: "0.5"
        memory: "50Mi"
  - name: pod-resources-demo-ctr-2
    image: fedora
    command: [ "sleep", "inf" ]
```

> In this scenario, the **request guarantees for *both* containers in the Pod will align with the pod-level requests** (1 CPU and 100 MiB in this example). Furthermore, both containers together **will not be able to use more resources than specified in the pod-level limits**, ensuring they cannot exceed a combined total of 200 MiB of memory and 1 core of CPU.

Reviewer: That's a lot of bold. Try to use it sparingly; eg just for the word "not".

> **Crucial Note on In-Place Resize:** It is important to understand that **in-place resize of pod-level resources is currently NOT supported**. Attempting to modify the pod-level resource limits or requests on a running Pod will result in an error. This feature focuses on the initial declaration and overall resource envelope at the Pod level. This is distinct from "In-Place Pod Resize", which allows dynamic adjustment of container resources within a *running* Pod without a restart and graduated to Beta in v1.33.

Reviewer: ```suggestion

## Limitations

First of all, [in-place resize](/docs/concepts/workloads/pods/#pod-update-and-replacement) of pod-level resources is **not supported** for Kubernetes v1.34 (or earlier).
Attempting to modify the _pod-level_ resource limits or requests on a running Pod results in an error:
the resize is rejected.
The v1.34 implementation of Pod level resources
focuses on allowing initial declaration of an overall resource envelope, that
applies to the **entire Pod**.
That is distinct from im-place pod resize, which
(despite what the name might suggest) allows you
to make dynamic adjustments to _container_ resource
requests and limits, within a *running* Pod, 
and potentially without a container restart.

In-place resizing is also not yet a stable feature; it graduated to Beta in the v1.33 release.
```

The new hyperlink is not an ideal target; we'll try to get the folks behind in-place resize to document it, but it is better to link there than not at all.

> #### **Getting Started and Providing Feedback**

Reviewer: ```suggestion
## Getting started, and ways to provide feedback
```

> Ready to explore Pod Level Resources? You'll need a Kubernetes cluster running version 1.34 or later. Remember to **enable the <code>PodLevelResources</code> feature gate** for your control plane and all nodes.

Reviewer: ```suggestion
Ready to explore Pod level resources? You'll need a Kubernetes cluster running version 1.34 or later. Remember to **enable the `PodLevelResources` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) across your control plane and all nodes.

If you are planning to upgrade to Kubernetes v1.34, I recommend that you do that first, and
then enable the new feature once the upgrade is complete.
```

---
layout: blog
title: "Kubernetes v1.34: Pod Level Resources Graduated to Beta"
date: 2025-xx-xx
draft: true
slug: kubernetes-v1-34-pod-level-resources
author: Dixita Narang (Google)
---

On behalf of the Kubernetes community, I am thrilled to announce that the Pod Level Resources feature has graduated to Beta in the Kubernetes v1.34 release! This significant milestone introduces a new layer of flexibility for defining and managing resource allocation for your Pods. This flexibility stems from the ability to specify CPU and memory resources directly at the Pod level, for the Pod as a whole, in addition to or instead of container-level specifications.

#### What are Pod Level Resources?

> Until recently, resource specifications that applied to Pods were primarily defined

Reviewer 2: primary presumes that there were a secondary way. Or you mean that some people didn't specify it at all?

> Until recently, resource specifications that applied to Pods were primarily defined

Author: May I mark this as resolved or should I remove "primarily"?

---
layout: blog
title: "Kubernetes v1.34: Pod Level Resources Graduated to Beta"
date: 2025-xx-xx
draft: true
slug: kubernetes-v1-34-pod-level-resources
author: Dixita Narang (Google)
---

On behalf of the Kubernetes community, I am thrilled to announce that the Pod Level Resources feature has graduated to Beta in the Kubernetes v1.34 release and is enabled by default! This significant milestone introduces a new layer of flexibility for defining and managing resource allocation for your Pods. This flexibility stems from the ability to specify CPU and memory resources for the Pod as a whole. Pod level resources can be combined with the container-level specifications to express the exact resource requirements and limits your application needs.

> #### What are Pod Level Resources?

Reviewer: ```suggestion
## Pod-level specification for resources
```
?

Reviewer: We do prefer headings in sentence case.

Until recently, resource specifications that applied to Pods were primarily defined
at the individual container level. While effective, this approach sometimes required
duplicating or meticulously calculating resource needs across multiple containers
within a single Pod. As a beta feature, Kubernetes allows you to specify the CPU,
memory and hugepages resources at the Pod-level. This means you can now define
resource requests and limits for an entire Pod, enabling easier resource sharing
without requiring granular, per-container management of these resources where
it's not needed.

> #### Why Does Pod Level Resources Matter?

Reviewer: ```suggestion
## Why does Podl-level specification matter?
```

Reviewer: We recommend sentence case headings. Also, this should not be a level 4 heading.

This feature enhances resource management in Kubernetes by offering *flexible resource management* at both the Pod and container levels.

* It provides a consolidated approach to resource declaration, reducing the need for

>   meticulous, per-container micromanagement, especially for Pods with multiple

Reviewer 2: ```suggestion
  meticulous, per-container management, especially for Pods with multiple
```

I feel like the word micromanagement has a negative feeling. And this is not micro in some cases - contaioners can be huge.

  containers. 
* Pod-level resources enable containers within a pod to share unused resoures

>   amongst themselves, promoting efficient utilization within the pod. For example, it prevents sidecar containers from becoming performance bottlenecks. Previously, a sidecar (e.g., a logging agent or service mesh proxy) hitting its individual CPU limit could be throttled and slow down the entire Pod, even if the main application container had plenty of spare CPU. With pod-level resources, the sidecar can borrow unused resources from the Pod's shared budget, ensuring smooth operation during traffic spikes.

Reviewer 2: ```suggestion
  amongst themselves, promoting efficient utilization within the pod. For example, it prevents sidecar containers from becoming performance bottlenecks. Previously, a sidecar (e.g., a service mesh proxy) hitting its individual CPU limit could be throttled and slow down the entire Pod, even if the main application container had plenty of spare CPU. With pod-level resources, the sidecar can borrow unused resources from the Pod's shared budget, ensuring smooth operation during traffic spikes.
```

Reviewer 2: this is not resolved. The paragraph emphasizes sharing, while I think it is more important to emphasize that the limit will be applied uniformely at the same time to both - sidecar and a main container

* When both pod-level and container-level resources are specified, pod-level
  requests and limits take precedence. This gives you – and cluster administrators -
  a powerful way to enforce overall resource boundaries for your Pods. 

  For scheduling, if a pod-level request is explicitly defined, the scheduler uses
  that specific value to find a suitable node, insteaf of the aggregated requests of
  the individual containers. At runtime, the pod-level limit acts as a hard ceiling
  for the combined resource usage of all containers. Crucially, this pod-level limit
  is the absolute enforcer; even if the sum of the individual container limits is
  higher, the total resource consumption can never exceed the pod-level limit.
* Pod-level resources are **prioritized** in influencing the Quality of Service (QoS) class of the Pod.
* For Pods running on Linux nodes, the Out-Of-Memory (OOM) score adjustment
  calculation considers both pod-level and container-level resources requests.
* Pod-level resources are **designed to be compatible with existing Kubernetes functionalities**, ensuring a smooth integration into your workflows.

> #### How to Use Pod Level Resources

Reviewer: ```suggestion
## How to specify resources for an entire Pod
```

Using `PodLevelResources` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/) requires
Kubernetes v1.34 or newer for all cluster components, including the control plane
and every node. This feature gate is in beta and enabled by default in v1.34.

> ##### Defining Pod-Level Resources

Reviewer: ```suggestion
### Example manifest
```

Reviewer: Make sure to, at minimum, use the correct heading levels and to write the headings in lower case.

    # The sum of the limits of all containers in the Pod cannot exceed these values.
    limits:
      cpu: "1" # The entire Pod cannot use more than 1 CPU core.
      memory: "200Mi" # The entire Pod cannot use more than 200 MiB of memory.
    # 'requests' specifies the minimum amount of resources guaranteed to the Pod.
    # This value is used by the Kubernetes scheduler to find a node with enough capacity.
    requests:
      cpu: "1" # The Pod is guaranteed 1 CPU core when scheduled.
      memory: "100Mi" # The Pod is guaranteed 100 MiB of memory when scheduled.
  containers:
  - name: main-app-container
    image: nginx
    ...
    # This container has no resource requests or limits specified.
  - name: auxiliary-container
    image: fedora
    command: ["sleep", "inf"]
    ...
    # This container has no resource requests or limits specified.
```

In this example, the `pod-resources-demo` Pod as a whole requests 1 CPU and 100 MiB of memory, and is limited to 1 CPU and 200 MiB of memory. The containers within will operate under these overall Pod-level constraints, as explained in the next section.

> #### Interaction with Container-Level Resources

Reviewer: ```suggestion
### Interaction with container-level resource requests or limits
```
?

Reviewer: Make sure to, at minimum, use the correct heading levels and to write the headings in lower case.

  - name: auxiliary-container
    image: fedora
    command: [ "sleep", "inf"]
    # This container has no resource requests or limits specified.
```

* Pod-Level Limits: The pod-level limits (cpu: "1", memory: "200Mi") establish an absolute boundary for the entire Pod. The sum of resources consumed by all its containers is enforced at this ceiling and cannot be surpassed.

* Resource Sharing and Bursting: Containers can dynamically borrow any unused capacity, allowing them to burst as needed, so long as the Pod's aggregate usage stays within the overall limit.

* Pod-Level Requests: The pod-level requests (cpu: "1", memory: "100Mi") serve as the foundational resource guarantee for the entire Pod. This value informs the scheduler's placement decision and represents the minimum resources the Pod can rely on during node-level contention.

* Container-Level Requests: Container-level requests create a priority system within
the Pod's guaranteed budget. Because main-app-container has an explicit request
(cpu: "0.5", memory: "50Mi"), it is given precedence for its share of resources
under resource pressure over the auxiliary-container, which has no
such explicit claim.

> #### Limitations

Reviewer: ```suggestion
## Limitations
```

  focuses on allowing initial declaration of an overall resource envelope, that
  applies to the **entire Pod**. That is distinct from in-place pod resize, which
  (despite what the name might suggest) allows you
  to make dynamic adjustments to _container_ resource
  requests and limits, within a *running* Pod, 
  and potentially without a container restart. In-place resizing is also not yet a
  stable feature; it graduated to Beta in the v1.33 release.

* Only CPU, memory, and hugepages resources can be specified at pod-level.

* Pod-level resources are not supported for Windows pods. If the Pod specification 
explicitly targets Windows (e.g., by setting spec.os.name: "windows"), the API
server will reject the Pod during the validation step. If the Pod is not explicitly
marked for Windows but is scheduled to a Windows node (e.g., via a nodeSelector),
the Kubelet on that Windows node will reject the Pod during its admission process. 

* The Topology Manager, Memory Manager and CPU Manager do not
  align pods and containers based on pod-level resources as these resource managers 
  don't currently support pod-level resources.

> #### Getting Started and Providing Feedback

Reviewer: ```suggestion
## Getting started and providing feedback
```

> Ready to explore Pod Level Resources? You'll need a Kubernetes cluster running version 1.34 or later. Remember to enable the `PodLevelResources` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) across your control plane and all nodes.

Reviewer: ```suggestion
Ready to explore the _Pod level resources_ feature?
You'll need a Kubernetes cluster running version 1.34 or later. Make sure to enable the `PodLevelResources` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
across your control plane and for all nodes.
```

---
layout: blog

> title: "Pod Level Resources Graduated to Beta"

Reviewer 3: ```suggestion
title: "Kubernetes v1.34: Pod Level Resources Graduated to Beta"
```

Reviewer: See suggestion

date: 2025-xx-xx
draft: true

> slug: pod-level-resources

Reviewer 3: ```suggestion
slug: kubernetes-v1-34-pod-level-resources
```
